WIP: tfvars files should be ignored during *-all commands if they have no source setting and no .tf file is present

jonahh-yeti opened the pull request:
Bit of a fresh GO programmer, but this change should do the trick.  The outstanding issue is that since the test data used is mostly just blank .tfvars files, any attempt to resolve these is now ignored so a ton of tests are failing.  Any advice on how to proceed or comments on the code would be welcome :)

Fix for issue https://github.com/gruntwork-io/terragrunt/issues/208

brikis98 commented on configstack/module.go:
```diff
@@ -83,7 +84,18 @@ func resolveTerraformModule(terragruntConfigPath string, terragruntOptions *opti
 
-	return &TerraformModule{Path: modulePath, Config: *terragruntConfig, TerragruntOptions: opts}, nil
+	if terragruntConfig.Terraform != nil {
+		s := []string{filepath.Dir(terragruntConfigPath), "*.tf"}
```
Shouldn't the check for `.tf` files happen whether or not a `terraform { ... }` block exists?

brikis98 replied:
Also, instead of using an array here and then joining it below, try:

```go
matches, err := filepath.Glob(filepath.Join(filepath.Dir(terragruntConfigPath), "*.tf"))
```

jonahh-yeti replied:
Yeah that's much cleaner, thanks

brikis98 commented on configstack/module.go:
```diff
@@ -83,7 +84,18 @@ func resolveTerraformModule(terragruntConfigPath string, terragruntOptions *opti
+		if err != nil {
+			return nil, err
+		}
+		if terragruntConfig.Terraform.Source != "" || matches != nil {
```
Reading the double-negative is a bit hard. Perhaps it would be simpler to invert this check?

```go
if terragruntConfig.Terraform.Source == "" && matches == nil {
  terragruntOptions.Logger.Printf("Module %s does not have an associated terraform config, sipping...", modulePath)
  return nil, nil
}
```

jonahh-yeti replied:
Looks like Go does use short-circuit evaluation on its Boolean operators, the new logic wraps the nil, nil return rather than the primary return and should look much cleaner!

jonahh-yeti commented:
Hey!  So, I made changes around your recommendations in the code.  The logic looks much cleaner overall now.  Also changed the tests so they don't all get caught by this addition, had to both edit the files in the test/fixture-modules/ directories and the expected values in the tests.  Finally, added a new test case and data in test/fixture-modules/module-l/ to check that a .tfvars file with no Terragrunt block and associated source, and no .tf file in the directory, will be skipped.  All tests are passing locally.  

Let me know what you think!

brikis98 commented on configstack/module.go:
```diff
@@ -83,6 +84,16 @@ func resolveTerraformModule(terragruntConfigPath string, terragruntOptions *opti
+		return nil, err
+	}
+	if (terragruntConfig.Terraform == nil || terragruntConfig.Terraform.Source == "") && matches == nil {
+		terragruntOptions.Logger.Printf("This module does not have an associated terraform config, sipping...")
```
s/sipping/skipping/. Also, include the module path in this output so we know which module is being skipped!

jonahh-yeti replied:
Fixed!  Can't believe I didn't catch the spelling error myself, and it'll include the directory path when it logs.

brikis98 commented:
Thank you, merging now!

brikis98 commented:
If the tests pass, the new binaries should show up here soon: https://github.com/gruntwork-io/terragrunt/releases/tag/v0.12.25
